mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,527 @@
+  if (use_previous) {
+    previous_policy.Action(action, state, time);
+  } else {
+    policy.Action(action, state, time);
+  }
+}
+
+// update policy via resampling
+void CEMPlanner::UpdateNominalPolicy(int horizon) {
+  // dimensions
+  int num_spline_points = candidate_policy[winner].num_spline_points;
+
+  // set time
+  double nominal_time = time;
+  double time_shift = mju_max(
+      (horizon - 1) * model->opt.timestep / (num_spline_points - 1), 1.0e-5);
+
+  // temp variables to help with averaging
+  std::vector<double> temp_avg(model->nu);  // holds avg for spline point t
+  // std::vector<double> temp_elite_actions(model->nu);
+  // std::array<std::array<double, nu>, n_elites> temp_elite_actions;
+  std::vector<std::vector<double>> temp_elite_actions;
+  temp_elite_actions.resize(n_elites, std::vector<double>(model->nu, 0.0));
+
+  // get spline points
+  for (int t = 0; t < num_spline_points; t++) {
+    times_scratch[t] = nominal_time;
+
+    // get the actions of the top n_elites policies and average them
+    // also update a variance parameter too
+    for (int i = 0; i < n_elites; i++) {
+      int index = trajectory_order[i];  // the (i+1)^th best trajectory [TODO] this segfaults!
+      candidate_policy[index].Action(DataAt(temp_elite_actions[i], 0), nullptr, nominal_time);
+      for (int k = 0; k < model->nu; k++) {
+        temp_avg[k] += temp_elite_actions[i][k] / num_spline_points;
+      }
+    }
+
+    // computing the sample variance of the control parameters
+    for (int k = 0; k < model->nu; k++) {
+      for (int i = 0; i < n_elites; i++) {
+        stdev[t * model->nu + k] += std::pow(temp_elite_actions[i][k] - temp_avg[k], 2) / (n_elites - 1);
+      }
+      stdev[t * model->nu + k] = std::sqrt(stdev[t * model->nu + k]);  // take the root
+    }
+
+    // assigning the averaged parameters to parameters_scratch
+    std::copy(temp_avg.begin(), temp_avg.end(), parameters_scratch.begin() + t * model->nu);
+    nominal_time += time_shift;
+  }
+
+  // update
+  {
+    const std::shared_lock<std::shared_mutex> lock(mtx_);
+    // parameters
+    policy.CopyParametersFrom(parameters_scratch, times_scratch);
+
+    // time power transformation
+    PowerSequence(policy.times.data(), time_shift,
+                  policy.times[0],
+                  policy.times[num_spline_points - 1],
+                  timestep_power, num_spline_points);
+  }
+}
+
+// add random noise to nominal policy
+void CEMPlanner::AddNoiseToPolicy(int i) {
+  // start timer
+  auto noise_start = std::chrono::steady_clock::now();
+
+  // dimensions
+  int num_spline_points = candidate_policy[i].num_spline_points;
+  int num_parameters = candidate_policy[i].num_parameters;
+
+  // sampling token
+  absl::BitGen gen_;
+
+  // shift index
+  int shift = i * (model->nu * kMaxTrajectoryHorizon);
+
+  // sample noise
+  for (int k = 0; k < num_parameters; k++) {
+    noise[k + shift] = absl::Gaussian<double>(gen_, 0.0, stdev[k]);
```

Comment:
note that the noise we add depends on `stdev`, which is computed when the nominal policy is updated.

Reply:
could very easily be an indexing bug here, need to check.